ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+**/
+SMMUV3_STREAM_TABLE_ENTRY *
+EFIAPI
+SmmuV3AllocateStreamTable (
+  IN SMMU_INFO    *SmmuInfo,
+  IN SMMU_CONFIG  *SmmuConfig,
+  OUT UINT32      *Log2Size,
+  OUT UINT32      *Size
+  )
+{
+  UINT32  MaxStreamId;
+  UINT32  SidMsb;
+  UINT32  Alignment;
+  UINTN   Pages;
+  VOID    *AllocatedAddress;
+
+  MaxStreamId      = SmmuConfig->Config.SmmuNode.SmmuIdMap.OutputBase + SmmuConfig->Config.SmmuNode.SmmuIdMap.NumIds;
+  SidMsb           = HighBitSet32 (MaxStreamId);
+  *Log2Size        = SidMsb + 1;
+  *Size            = SMMUV3_LINEAR_STREAM_TABLE_SIZE_FROM_LOG2 (*Log2Size);
+  *Size            = ROUND_UP (*Size, EFI_PAGE_SIZE);
+  Alignment        = ALIGN_UP_BY (*Size, EFI_PAGE_SIZE);
+  Pages            = EFI_SIZE_TO_PAGES (*Size);
+  AllocatedAddress = AllocateAlignedPages (Pages, Alignment);
+
+  ASSERT (AllocatedAddress != NULL);
+  ZeroMem (AllocatedAddress, *Size);
+  return (SMMUV3_STREAM_TABLE_ENTRY *)AllocatedAddress;
+}
+
+/**
+  Free the allocated stream table for SMMUv3.
+
+  @param [in] StreamTablePtr  Pointer to the stream table entry.
+  @param [in] Size            Size of the stream table.
+**/
+VOID
+EFIAPI
+SmmuV3FreeStreamTable (
+  IN SMMUV3_STREAM_TABLE_ENTRY  *StreamTablePtr,
+  IN UINT32                     Size
+  )
+{
+  UINTN  Pages;
+
+  Pages = EFI_SIZE_TO_PAGES (Size);
+  FreeAlignedPages ((VOID *)StreamTablePtr, Pages);
+}
+
+/**
+  Configure the SMMUv3 based on the provided configuration per the SmmuV3 specification.
+
+  @param [in] SmmuInfo    Pointer to the SMMU_INFO structure.
+  @param [in] SmmuConfig  Pointer to the SMMU configuration.
+
+  @retval EFI_SUCCESS     Success.
+  @retval Others          Failure.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3Configure (
+  IN SMMU_INFO    *SmmuInfo,
+  IN SMMU_CONFIG  *SmmuConfig
+  )
+{
+  EFI_STATUS                 Status;
+  UINT32                     STLog2Size;
+  UINT32                     STSize;
+  UINT32                     CommandQueueLog2Size;
+  UINT32                     EventQueueLog2Size;
+  UINT8                      ReadWriteAllocationHint;
+  SMMUV3_STRTAB_BASE         StrTabBase;
+  SMMUV3_STRTAB_BASE_CFG     StrTabBaseCfg;
+  SMMUV3_STREAM_TABLE_ENTRY  *StreamTablePtr;
+  SMMUV3_CMDQ_BASE           CommandQueueBase;
+  SMMUV3_EVENTQ_BASE         EventQueueBase;
+  SMMUV3_STREAM_TABLE_ENTRY  TemplateStreamEntry;
+  SMMUV3_CR0                 Cr0;
+  SMMUV3_CR1                 Cr1;
+  SMMUV3_CR2                 Cr2;
+  SMMUV3_IDR0                Idr0;
+  SMMUV3_GERROR              GError;
+  SMMUV3_CMD_GENERIC         Command;
+  VOID                       *CommandQueue;
+  VOID                       *EventQueue;
+
+  if ((SmmuConfig->Config.SmmuNode.SmmuNode.Flags & EFI_ACPI_IORT_SMMUv3_FLAG_COHAC_OVERRIDE) != 0) {
+    ReadWriteAllocationHint = 0x1;
+  } else {
+    ReadWriteAllocationHint = 0x0;
+  }
+
+  GError.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_GERROR);
+  ASSERT (GError.AsUINT32 == 0);

```

Comment:
Should the configure fail if this is non-zero? We should handle this gracefully.

Reply:
done